sdk/Pulumi/Resources/CustomTimeouts.cs

```diff
@@ -31,5 +31,128 @@ public sealed class CustomTimeouts
+        {
+            if (timeSpan == null)
+                return "";
+
+            // This will eventually be parsed by go's ParseDuration function here:
+            // https://github.com/pulumi/pulumi/blob/06d4dde8898b2a0de2c3c7ff8e45f97495b89d82/pkg/resource/deploy/source_eval.go#L967
+            //
+            // So we generate a legal duration as allowed by
+            // https://golang.org/pkg/time/#ParseDuration.
+            //
+            // Simply put, we simply convert our ticks to the integral number of nanoseconds
+            // corresponding to it.  Since each tick is 100ns, this can trivially be done just by
+            // appending "00" to it.
+            return timeSpan.Value.Ticks + "00ns";
+        }
+
+        internal Pulumirpc.RegisterResourceRequest.Types.CustomTimeouts Serialize()
+        {
+            return new Pulumirpc.RegisterResourceRequest.Types.CustomTimeouts
+            {
+                Create = TimeoutString(Create),
+                Update = TimeoutString(Update),
+                Delete = TimeoutString(Delete),
+            };
+        }
+
+        internal static CustomTimeouts Deserialize(Pulumirpc.RegisterResourceRequest.Types.CustomTimeouts customTimeouts)
+        {
+            Func<string, TimeSpan?> parse = (string s) =>
+            {
+                if (s == null || s == "")
+                {
+                    return null;
+                }
+
+                // A duration string is a possibly signed sequence of decimal numbers, each with optional
+                // fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time units are "ns",
+                // "us" (or "µs"), "ms", "s", "m", "h".
+
+                var orig = s;
+                var neg = false;
+                if (s[0] == '-' || s[0] == '+')
+                {
+                    neg = s[0] == '-';
+                    s = s.Substring(1);
+                }
+                if (s == "0")
+                {
+                    return TimeSpan.Zero;
+                }
+                if (s == "")
+                {
+                    throw new ArgumentException("invalid duration " + s);
+                }
+                var duration = TimeSpan.Zero;
+                while (s != "")
+                {
+                    // find the next timeunit
+                    var i = 0;
+                    while (i < s.Length && (('0' <= s[i] && s[i] <= '9') || s[i] == '.'))
+                    {
+                        i++;
+                    }
+                    // parse the number
+                    var v = double.Parse(s.Substring(0, i));
+                    // parse the unit
+                    s = s.Substring(i);
+                    if (s == "")
+                    {
+                        throw new ArgumentException("missing unit in duration " + orig);
+                    }
+                    if (s.StartsWith("µs") || s.StartsWith("us"))
```

Comment:
Operating on `ReadOnlySpan<char>` will avoid this, but just pointing out that `string.StartsWith` does a culture-sensitive comparison by default. If we're keeping this a string, we should use `s.StartsWith("µs", StringComparison.Ordinal)`.